### PR TITLE
Filter Notification messages exclusively to Gravity PDF pages.

### LINKED ISSUE #1345: After  new PDF form submission, message should show along side other add-on messages.
**Description**

**Step To Reproduce**
1. Go to Install an add-on that requires other dependencies .
2. Click on 'Form -> Settings->PDF->Add new'
3. Submit a new PDF template

**Expected behavior**
Success message should show along side with other  alerts.
image

**Screenshots**
image

**Additional context**
I believe we already did some update regarding this. AFAIK, we prepend our message to `.gform-settings__content`.

### PR DESCRIPTION
**Pull request by jestonihpi**
Create a separate method that returns our `errors` and `notices`  messages exclusively to Gravity PDF pages.

NOTE: This may need some improvements on how to properly initialize the Helper_Notice class exclusively to Gravity PDF pages as well.
## Testing instructions
Install Gravity Forms Akismet Add-On. Its default error message should not display on PDF settings.

## Checklist:
- [x] I've tested the code.
- [ ] My code is easy to read, follow, and understand
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation / docblocks.

### CODE REVIEW

File: src/Helper/Helper_Notices.php [hunk 1]
```diff
@@ -210,4 +212,32 @@ private function html( $text, $class = 'updated' ) {
...
+	 *
+	 * @param array $messages  The message to be displayed
+	 *
+	 * @return array $messages
+	 *
+	 * @since 6.3
+	 */
+	public function filter_messages( $messages ) {
+		global $gfpdf;
+		$misc = new Helper_Misc( $gfpdf->log, $gfpdf->gform, $gfpdf->data );
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: `\GPDFAPI::get_misc_class()` would be a better option.

**Review by jakejackson1 — CHANGES_REQUESTED** [review]
This isn't the best way to tackle this. 

I recommend removing L91-95 so that the Gravity PDF notices don't get mixed up with the Gravity Forms notices.

I'd then create an individual method for the hooks on L62 and L63 and check if it's a Gravity PDF page. If it is a GPDF page, return the queued Gravity PDF notice or error message.

**Comment by jakejackson1** [comment]
Also, missing unit tests.

File: src/Helper/Helper_Notices.php [hunk 2]
```diff
@@ -210,4 +203,37 @@ private function html( $text, $class = 'updated' ) {
 		</div>
 		<?php
 	}
+
+	/**
+	 * If the current page belongs to Gravity PDF, add these hooks.
+	 *
+	 * @since 6.3
+	 */
+	public function maybe_gravitypdf_page() {
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: This is a poorly-named method. Perhaps `maybe_remove_non_pdf_notices`?

File: src/Helper/Helper_Notices.php [hunk 3]
```diff
@@ -210,4 +203,37 @@ private function html( $text, $class = 'updated' ) {
 		</div>
 		<?php
 	}
+
+	/**
+	 * If the current page belongs to Gravity PDF, add these hooks.
+	 *
+	 * @since 6.3
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: This should be 6.4

File: src/Helper/Helper_Notices.php [hunk 4]
```diff
@@ -210,4 +203,37 @@ private function html( $text, $class = 'updated' ) {
...
+	public function maybe_gravitypdf_page() {
+		if ( \GPDFAPI::get_misc_class()->is_gfpdf_page() ) {
+			add_action( 'gform_admin_messages ', [ $this, 'remove_non_gravity_pdf_messages' ] );
+			add_action( 'gform_admin_error_messages', [ $this, 'remove_non_gravity_pdf_messages' ] );
+		}
+	}
+
+	/**
+	 * Remove all Non-Gravity PDF messages.
+	 * Due GFCommon::display_admin_message that only display one message at time.
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: I have zero clue what this means. Add appropriate context.

File: src/Helper/Helper_Notices.php [hunk 5]
```diff
@@ -210,4 +203,37 @@ private function html( $text, $class = 'updated' ) {
...
+	 *
+	 * @param array $messages  The message to be displayed
+	 *
+	 * @return array $messages
+	 *
+	 * @since 6.3
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: 6.4

File: src/Helper/Helper_Notices.php [hunk 6]
```diff
@@ -210,4 +203,37 @@ private function html( $text, $class = 'updated' ) {
...
 	}
+
+	/**
+	 * If the current page belongs to Gravity PDF, add these hooks.
+	 *
+	 * @since 6.3
+	 */
+	public function maybe_gravitypdf_page() {
+		if ( \GPDFAPI::get_misc_class()->is_gfpdf_page() ) {
+			add_action( 'gform_admin_messages ', [ $this, 'remove_non_gravity_pdf_messages' ] );
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: Remove space after `_messages` and `'`

File: src/Helper/Helper_Notices.php [hunk 7]
```diff
@@ -112,7 +106,6 @@ public function add_notice( $notice, $class = '' ) {
 	 * @since    4.0
 	 */
 	public function add_error( $error, $class = '' ) {
-
 		if ( GFForms::is_gravity_page() ) {
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: We did we remove this check from the notice method and not the error as well?

File: src/Helper/Helper_Notices.php [hunk 8]
```diff
@@ -210,4 +203,37 @@ private function html( $text, $class = 'updated' ) {
...
+	 * Due GFCommon::display_admin_message that only display one message at time.
+	 *
+	 * @param array $messages  The message to be displayed
+	 *
+	 * @return array $messages
+	 *
+	 * @since 6.3
+	 */
+	public function remove_non_gravity_pdf_messages( $messages ) {
+		foreach ( $messages as $i => $message ) {
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: The idea is to disregard `$messages` on Gravity PDF-related pages and return `$this->notices` or `$this->errors` instead, thereby ensuring that non-Gravity PDF notices/errors won't show up on Gravity PDF pages.

**Review by jakejackson1 — CHANGES_REQUESTED** [review]
The PR description and title should be updated appropriately.

File: src/Helper/Helper_Notices.php [hunk 9]
```diff
@@ -225,4 +213,51 @@ private function html( $text, $class = 'updated' ) {
...
+	 * @return array $this->errors
+	 *
+	 * @since 6.4
+	 */
+	public function return_pdf_error_messages( array $messages ): array {
+		if ( $this->has_error() ) {
+			return $this->errors;
+		}
+
+		return [];
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: Let's reverse this so that if there are no errors we return an empty array, but if there are errors we return the errors. It makes the code more readable by putting the negative condition inside the IF statement and the positive condition in the top-level flow. The same goes for the notices.

File: tests/phpunit/unit-tests/test-notices.php [hunk 10]
```diff
@@ -159,4 +159,13 @@ public function test_html_notice() {
 
 		$this->assertNotFalse( strpos( $html, $form ) );
 	}
+
+	/* Test empty $this->errors and $this->notices */
+	public function test_gform_admin_messages_hooks() {
```
Comment (jakejackson1) [inline tests/phpunit/unit-tests/test-notices.php:167]: I think we need to enhance the tests here to check specifically that the registered errors/notices are returned on Gravity PDF pages that run the `gform_admin_error_messages` or `gform_admin_messages` filters, but are NOT returned if the pages are not Gravity PDF ones.

File: src/Helper/Helper_Notices.php [hunk 11]
```diff
@@ -225,4 +213,51 @@ private function html( $text, $class = 'updated' ) {
...
+	/**
+	 * Return Gravity PDF error messages and disregard the existing messages.
+	 *
+	 * @param array $messages The message to be displayed
+	 *
+	 * @return array $this->errors
+	 *
+	 * @since 6.4
+	 */
+	public function return_pdf_error_messages( array $messages ): array {
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: A better name for this method would be `override_gravityforms_admin_error_messages()`

File: src/Helper/Helper_Notices.php [hunk 12]
```diff
@@ -225,4 +213,51 @@ private function html( $text, $class = 'updated' ) {
...
+	/**
+	 * Return Gravity PDF notice messages and disregard the existing messages.
+	 *
+	 * @param array $messages The message to be displayed
+	 *
+	 * @return array $this->notices
+	 *
+	 * @since 6.4
+	 */
+	public function return_pdf_notice_messages( array $messages ): array {
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: A better name for this method would be `override_gravityforms_admin_messages()`

File: tests/phpunit/unit-tests/test-notices.php (hunk 10 above)
Comment (jakejackson1) [reply tests/phpunit/unit-tests/test-notices.php:167]: You can use `apply_filters()` to trigger the correct filters, but you'll need to override the globals to mock Gravity PDF vs non-Gravity PDF pages.

File: src/Helper/Helper_Notices.php [hunk 13]
```diff
@@ -225,4 +213,51 @@ private function html( $text, $class = 'updated' ) {
 
 		remove_filter( 'wp_kses_allowed_html', $allow_form_elements );
 	}
+
+	/**
+	 * If the current page belongs to Gravity PDF, add these hooks.
+	 *
+	 * @since 6.4
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: As 6.4 is now released, this should be updated to 6.5

File: src/Helper/Helper_Notices.php [hunk 14]
```diff
@@ -225,4 +213,51 @@ private function html( $text, $class = 'updated' ) {
...
+	}
+
+	/**
+	 * Return Gravity PDF error messages and disregard the existing messages.
+	 *
+	 * @param array $messages The message to be displayed
+	 *
+	 * @return array $this->errors
+	 *
+	 * @since 6.4
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: As 6.4 is now released, this should be updated to 6.5

File: src/Helper/Helper_Notices.php [hunk 15]
```diff
@@ -225,4 +213,51 @@ private function html( $text, $class = 'updated' ) {
...
+	}
+
+	/**
+	 * Return Gravity PDF notice messages and disregard the existing messages.
+	 *
+	 * @param array $messages The message to be displayed
+	 *
+	 * @return array $this->notices
+	 *
+	 * @since 6.4
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: As 6.4 is now released, this should be updated to 6.5

**Review by jakejackson1 — CHANGES_REQUESTED** [review]
@jestonihpi this is looking a lot better since last time I checked it out. The unit tests are lacking though and I'd like to see more tests to verify this functions as expected.

File: src/Helper/Helper_Notices.php [hunk 16]
```diff
@@ -59,6 +59,20 @@ public function init() {
...
+	}
+
+	/**
+	 * If the current page belongs to Gravity PDF, add these hooks.
+	 * Remove these hooks if the current page does not belong to Gravity PDF.
+	 *
+	 * @since 6.5
+	 */
+	public function maybe_remove_non_pdf_messages() {
+		if ( \GPDFAPI::get_misc_class()->is_gfpdf_page() ) {
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: Let's reverse this condition and return early if not a GF page. It makes the flow easier to follow as you can have the main logic below it.

**Review by jakejackson1 — CHANGES_REQUESTED** [review]
Per the Codecov alert on L241 and L258, we also need tests that verify an empty array is returned when on GPDF pages.

**Comment by jestonihpi** [comment]
Hi @jakejackson1 , can you check on this PR as well? Thanks!

File: src/Helper/Helper_Notices.php [hunk 17]
```diff
@@ -225,4 +228,38 @@ private function html( $text, $class = 'updated' ) {
...
+	 *
+	 * @return array $this->notices
+	 *
+	 * @since 6.5
+	 */
+	public function override_gravityforms_admin_messages( $messages ): array {
+		if ( ! $this->has_notice() ) {
+			return [];
+		}
+		return $this->notices;
```
Comment (jakejackson1) [inline src/Helper/Helper_Notices.php]: @jestonihpi for consistency, add a new line before the return statement, like in the method above.

**Review by jakejackson1 — CHANGES_REQUESTED** [review]
@jestonihpi just a minor change for consistency and then this should be ready to go I think.

**Comment by jestonihpi** [comment]
Copy @jakejackson1. Thanks!

**Comment by jestonihpi** [comment]
Hey @jakejackson1 , this PR is now ready for your review.

**Comment by jakejackson1** [comment]
@jestonihpi I found problems with the notices being displayed when any Gravity PDF extensions are also activated on the site. This is because each extension creates a new Notices object on initialisation and they all add their own independent `maybe_remove_non_pdf_messages` hook (and subsequent override hooks). 

The exact flow when two Gravity PDF extensions are activated and the Core plugin saves the PDF settings to display the "Successful" message is:

1. Core Registers Notices class
1. Previewer Registers Notices class
1. Bulk Generator Registers Notices class
1. Core Registers Override Hooks
1. Previewer Registers Override Hooks
1. Bulk Generator Registers Override Hooks
1. Core Override Triggers and returns the "Success" message
1. Previewer Override Triggers and returns empty array
1. Bulk Generator Override Triggers and returns empty array 
1. Gravity Forms displays no notices

I think the best way to handle this is to split up the `gform_admin_messages` and `gform_admin_error_messages` actions we register so that we empty any existing messages at priority 999 (`return []` and add any Gravity PDF messages at priority 1000 `return array_merge( $messages, $this->notices )`.

This will ensure Gravity PDF Core and Extensions can still have unique Notices classes, non-Gravity PDF notices are removed on Gravity PDF pages, and each plugin won't override each other's messages. 

Does that make sense?

**Comment by jestonihpi** [comment]
None for now @jakejackson1 , but I will keep you posted for further questions.

**Comment by jestonihpi** [comment]
Hi @jakejackson1 , I have updated the unit tests. I have separated the non-filtered test and filtered. So with this, non-filtered should only reuse the notice object in the init, while the filtered test needs to re initialize starting from init.

**Comment by jestonihpi** [comment]
@jakejackson1 , converting this PR to draft for now. Its not working as expected. 
image

**Comment by jakejackson1** [comment]
@jestonihpi I reverted your last change. I believe the Core Font installer / registered actions inside of `Controller_Actions` / certain folder permissions make use of the notice class to output messages to the end user. These notices can be shown outside Gravity PDF pages and this handles that use-case. 

I also committed a tweak that should handle the GF-messages registered outside the `gform_admin_messages` processing (needs further testing).

**Comment by jestonihpi** [comment]
Copy @jakejackson1.
